scib/metrics/morans_i.py

```diff
@@ -0,0 +1,149 @@
+            print(
+                "Using subset (%i) of hvgs (%i) that are not constant in the data"
+                % (len(hvgs_used), len(hvgs))
+            )
+        # Compute Moran's I
+        morans_i = sc.metrics.morans_i(data[:, hvgs_used])
+        morans_i = pd.Series(morans_i, index=hvgs_used)
+        # This should not happen, just in case
+        if (morans_i > 1).any() or (morans_i < -1).any():
+            raise ValueError(
+                "Problems in computing Moran's I, value not between -1 and 1"
+            )
+        data.var["morans_i"] = morans_i
+
+    if compare_pre:
+        # Get per sample connectivities
+        adatas_sample = dict()
+        for sample in adata_pre.obs[batch_key].unique():
+            # Subset only to cells from 1 sample
+            adata_sample = adata_pre[adata_pre.obs[batch_key] == sample, :].copy()
+            # Compute HVG for each sample
+            sc.pp.highly_variable_genes(
+                adata_sample,
+                flavor="cell_ranger",
+                batch_key="study_sample",
+                n_top_genes=2000,
+            )
+            adata_sample_scl = adata_sample.copy()
+            # PP each sample: scaling, pca, neighbours
+            sc.pp.scale(adata_sample_scl, max_value=10)
+            sc.pp.pca(
+                adata_sample_scl,
+                n_comps=15,
+                use_highly_variable=True,
+                svd_solver="arpack",
+            )
+            sc.pp.neighbors(adata_sample_scl, n_pcs=15)
+            # Add computed info back to unscaled data
+            adata_sample.uns["neighbors"] = adata_sample_scl.uns["neighbors"]
+            adata_sample.obsp["connectivities"] = adata_sample_scl.obsp[
+                "connectivities"
+            ]
+            adata_sample.obsp["distances"] = adata_sample_scl.obsp["distances"]
+            adata_sample.uns["pca"] = adata_sample_scl.uns["pca"]
+            adata_sample.obsm["X_pca"] = adata_sample_scl.obsm["X_pca"]
+            # Save adata of sample
+            adatas_sample[sample] = adata_sample
+        del adata_sample
+        del adata_sample_scl
+
+        # Compute Moran's I across samples and on integrated data
+        for data in list(adatas_sample.values()) + [adata_post]:
+            compute_mi(data, hvgs)
+
+        # Compute differences in Moran's I
+        # Table of Moran's I-s across batches
+        batch_mis = []
+        for sample, data in adatas_sample.items():
+            mi = data.var["morans_i"]
+            mi.name = sample
+            batch_mis.append(mi)
+        batch_mis = pd.concat(batch_mis, axis=1)
+        # Difference with integrated data
+        mi_diffs = adata_post.var["morans_i"] - batch_mis.max(axis=1)
+        avg_mi_diff = mi_diffs.mean()
+
+        # Rescale so that it is between [0,1] where 1 is better
+        # Moran's I will be in [-1,1] and thus difference can be in [-2,2]
+        if rescale:
+            res = (avg_mi_diff + 2) / 4
```

Comment:
Another note on rescaling. Here it seems that a difference of 2 (perfect auto-correlation of markers in the integrated data but max per-batch auto-correlation is -1) is optimal. I'm not sure this should be the case. Wouldn't the idea be to detect the same spatial correlation structure in a single batch as across all of them? I guess you want a `1 - (avg_mi_diff + 2) / 4` here, no? 

Also, should a Moran's I of -1 and 0 be distinguished for the comparison at all? Both mean no biologically meaningful information is encoded, no?

Reply:
-1 means perfectly dispersed, if we are looking for a clustered signal we could do `max(0, score)` to have the score in `[0,1]` from the beginning - what do you think? 

Wrt to the `1 - ... `, I totally agree. Minimal distances should yield a high score.

Reply:
EDIT: I changed it such as to only consider "bad" differences, please check in #304